ccl/streamingccl/streamingest: TestTenantStreaming failed

ccl/streamingccl/streamingest.TestTenantStreaming [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5685691?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5685691?buildTab=artifacts#/) on master @ [88789feea153b61b379c0240a07482bdfaf0c932](https://github.com/cockroachdb/cockroach/commits/88789feea153b61b379c0240a07482bdfaf0c932):


```
=== RUN   TestTenantStreaming
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/045fecfe9e18c9da6f776a53d446f2c5/logTestTenantStreaming2237887931
    test_log_scope.go:80: use -show-logs to present logs inline
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/tenant-streaming
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantStreaming.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-17431

```
=== RUN   TestTenantStreaming
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/045fecfe9e18c9da6f776a53d446f2c5/logTestTenantStreaming2237887931
    test_log_scope.go:80: use -show-logs to present logs inline
panic: use of Span after Finish. Span: ingestfntr. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

goroutine 1617744 [running]:
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0xc002c5f180)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:186 +0x279
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Tracer).startSpanGeneric(0xc0138b2a50, {0x72291c0, 0xc019ebbe40}, {0x68e4eb2, 0x1d}, {{0x0}, 0x0, {0x0, 0x0, {{0x0, ...}, ...}, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/util/tracing/tracer.go:1207 +0x997
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Tracer).StartSpanCtx(0xc0138b2a50, {0x72291c0, 0xc019ebbe40}, {0x68e4eb2, 0x1d}, {0xc010cbb760, 0x1, 0x1})
	github.com/cockroachdb/cockroach/pkg/util/tracing/tracer.go:1062 +0x1a7
github.com/cockroachdb/cockroach/pkg/util/tracing.ChildSpan({0x72291c0, 0xc019ebbe40}, {0x68e4eb2, 0x1d})
	github.com/cockroachdb/cockroach/pkg/util/tracing/tracer.go:1577 +0x145
github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamclient.(*partitionedStreamClient).Heartbeat(0xc00b311080, {0x72291c0, 0xc019ebbe40}, 0xac93aa873318001, {0x16ffc388c7d7c42a, 0x0, 0x0})
	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go:83 +0xb7
github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*heartbeatSender).maybeHeartbeat(0xc01980f880, {0x72291c0, 0xc019ebbe40}, {0x16ffc388c7d7c42a, 0x0, 0x0})
	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go:180 +0x250
github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*heartbeatSender).startHeartbeatLoop.func1.1()
	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go:207 +0x41b
github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*heartbeatSender).startHeartbeatLoop.func1({0x72291c0, 0xc019ebbe40})
	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go:227 +0xb5
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:169 +0x52
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:74 +0xb4
created by golang.org/x/sync/errgroup.(*Group).Go
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:71 +0xdd
I220708 05:29:42.017167 1 (gostd) testmain.go:90  [-] 1  Test //pkg/ccl/streamingccl/streamingest:streamingest_test exited with error code 2
```

--- a/pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go
+++ b/pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go
@@ -125,6 +125,10 @@ func newStreamIngestionFrontierProcessor(
 		nil, /* memMonitor */
 		execinfra.ProcStateOpts{
 			InputsToDrain: []execinfra.RowSource{sf.input},
+			TrailingMetaCallback: func() []execinfrapb.ProducerMetadata {
+				sf.close()
+				return nil
+			},
 		},
 	); err != nil {
 		return nil, err
@@ -146,8 +150,8 @@ type heartbeatSender struct {
 	flowCtx         *execinfra.FlowCtx
 	// cg runs the heartbeatSender thread.
 	cg ctxgroup.Group
-	// Send signal to stopChan to stop heartbeat sender.
-	stopChan chan struct{}
+	// cancel stops heartbeat sender.
+	cancel func()
 	// heartbeatSender closes this channel when it stops.
 	stoppedChan chan struct{}
 }
@@ -164,7 +168,7 @@ func newHeartbeatSender(
 		streamID:        streaming.StreamID(spec.StreamID),
 		flowCtx:         flowCtx,
 		frontierUpdates: make(chan hlc.Timestamp),
-		stopChan:        make(chan struct{}),
+		cancel:          func() {},
 		stoppedChan:     make(chan struct{}),
 	}, nil
 }
@@ -182,6 +186,8 @@ func (h *heartbeatSender) maybeHeartbeat(
 }
 
 func (h *heartbeatSender) startHeartbeatLoop(ctx context.Context) {
+	ctx, cancel := context.WithCancel(ctx)
+	h.cancel = cancel
 	h.cg = ctxgroup.WithContext(ctx)
 	h.cg.GoCtx(func(ctx context.Context) error {
 		sendHeartbeats := func() error {
@@ -196,8 +202,6 @@ func (h *heartbeatSender) startHeartbeatLoop(ctx context.Context) {
 				select {
 				case <-ctx.Done():
 					return ctx.Err()
-				case <-h.stopChan:
-					return nil
 				case <-timer.C:
 					timer.Reset(streamingccl.StreamReplicationConsumerHeartbeatFrequency.
 						Get(&h.flowCtx.EvalCtx.Settings.SV))
@@ -231,15 +235,20 @@ func (h *heartbeatSender) startHeartbeatLoop(ctx context.Context) {
 }
 
 // Stop the heartbeat loop and returns any error at time of heartbeatSender's exit.
-// Should be called at most once.
+// Can be called multiple times.
 func (h *heartbeatSender) stop() error {
-	close(h.stopChan) // Panic if closed multiple times
-	return h.cg.Wait()
+	h.cancel()
+	return h.wait()
 }
 
 // Wait for heartbeatSender to be stopped and returns any error.
-func (h *heartbeatSender) err() error {
-	return h.cg.Wait()
+func (h *heartbeatSender) wait() error {
+	err := h.cg.Wait()
+	// We expect to see context cancelled when shutting down.
+	if errors.Is(err, context.Canceled) {
+		return nil
+	}
+	return err
 }
 
 // Start is part of the RowSource interface.
@@ -305,8 +314,10 @@ func (sf *streamIngestionFrontier) Next() (
 			// If heartbeatSender has error, it means remote has error, we want to
 			// stop the processor.
 		case <-sf.heartbeatSender.stoppedChan:
-			err := sf.heartbeatSender.err()
-			log.Warningf(sf.Ctx, "heartbeat sender has stopped with error: %s", err)
+			err := sf.heartbeatSender.wait()
+			if err != nil {
+				log.Errorf(sf.Ctx, "heartbeat sender exited with error: %s", err)
+			}
 			sf.MoveToDraining(err)
 			return nil, sf.DrainHelper()
 		}
@@ -314,14 +325,18 @@ func (sf *streamIngestionFrontier) Next() (
 	return nil, sf.DrainHelper()
 }
 
-// ConsumerClosed is part of the RowSource interface.
-func (sf *streamIngestionFrontier) ConsumerClosed() {
+func (sf *streamIngestionFrontier) close() {
+	if err := sf.heartbeatSender.stop(); err != nil {
+		log.Errorf(sf.Ctx, "heartbeat sender exited with error: %s", err)
+	}
 	if sf.InternalClose() {
-		if err := sf.heartbeatSender.stop(); err != nil {
-			log.Errorf(sf.Ctx, "heartbeatSender exited with error: %s", err.Error())
-		}
 		sf.metrics.RunningCount.Dec(1)
 	}
+}
+
+// ConsumerClosed is part of the RowSource interface.
+func (sf *streamIngestionFrontier) ConsumerClosed() {
+	sf.close()
 }
 
 // decodeResolvedSpans decodes an encoded datum of jobspb.ResolvedSpans into a

